### ISSUE TITLE
Retain orphaned blocks within archive database

### ISSUE THREAD

**Issue by benjaminion** [Consensys-Incorporated/teku#2853]
Change request from @mcdee via Discord.

Currently, Teku aggresively prunes orphaned blocks from the database, even in ARCHIVE mode. For data analytics purposes it would be useful to retain orphaned blocks and make them accessible via the API in ARCHIVE mode.

**Comment by benjaminion** [comment]
Noting so I don't forget that we now have a second request for this feature. Also that the API should indicate whether a block was canonical or orphaned - I don't know whether this is possible within the constraints of the current API.

**Comment by mcdee** [comment]
Regarding stating the canonical state of a block: it probably falls outside of the purview of the API to provide that info in the standard API call.  The API tries to return ETH2 structs where possible, and canonical state is somewhat another layer on top of the block information rather than an inherent property of it.

**Comment by rolfyone** [comment]
There should be a release tomorrow containing this feature behind the dev cli flag `--Xdata-storage-non-canonical-blocks-enabled`
and fetching block by root for orphaned blocks should then also work.

If you'd like to take a look and see if this is adequate for your use-case @mcdee ?  Historic orphaned blocks won't be able to be fetched, but new orphaned blocks should start getting stored, and you'd have to fetch by block root...

**Comment by mcdee** [comment]
Will fetching by slot also work?

Also, curious as to if the node will store historical orphaned blocks with this mode enabled, if these blocks are delivered to the node?  Wondering if I can find a way to back-fill the historic orphaned blocks by sending them over p2p.

I haven't checked the code so apologies if this is already the case, but does it make sense for this flag to be enabled by default in archive mode?

**Comment by rolfyone** [comment]
Fetch by slot won't currently work for orphaned blocks, is that functionality that you need?

Historical block fetching is only canonical, I'm not sure it'd be easy to sync non canonical slots through that process currently.

If the block was gossiped to the node, the node would then see it's non canonical during finalization and store it as a non canonical block... I've seen it working on current blocks, but nothing currently pushes non canonical blocks around as far as I'm aware. I added a debug for this, because its fairly rare on mainnet to even see a non canonical block ( a few a day)

So basically while the orphaned block is in non finalized state, it comes from hot storage (used to do that) and would be fetchable by slot or by root, because it's heavily indexed. Without the dev flag, it will be deleted during the move of data to finalized storage.

With the new flag, It will get stored and indexed by root, so retrieval by block root is cheap. Retrieval by slot hasn't currently been added, theres options but neither are great, I might actually need a separate endpoint that does a full scan on the new column.

**Comment by mcdee** [comment]
> Fetch by slot won't currently work for orphaned blocks, is that functionality that you need?

Not sure how it would be possible to discover orphaned blocks without the ability to fetch by slot.  Is there some other way in which these could be found?

**Comment by rolfyone** [comment]
I'm really not sure actually, I might tag @ajsutton and see if he's got any ideas...

**Comment by ajsutton** [comment]
> Not sure how it would be possible to discover orphaned blocks without the ability to fetch by slot. Is there some other way in which these could be found?

So the details depend a fair bit on use case.  Every block, canonical or not will be sent to the event stream which has the block hash.  But I suspect if you can catch the block hash from there then you'd get it before the slot was finalised and you'd have always been able to get it.

If the node is offline when the block is published it will probably never get it and definitely won't ever get it if it hasn't heard about it by the time the slot is finalized.

So I'm guessing that you're thinking in terms of something like having a node that's been consistently online but then something like chaind has just restarted and is backfilling data for certain epochs.  In that case yeah you'd probably have to walk through each slot between the last block you have and the current chain head requesting canonical and non-canonical at each slot.

The other option would be to request based on the parent_root which oddly the standard API does support (though Teku's never implemented it because it seems weird). In that case you'd have to get each canonical block and then use it's hash as the parent_root to see if it has any non-canonical descendants and follow each chain you find.  Get by slot probably makes more sense.

We'll have to add a teku specific API to be able to get non-canonical blocks by slot and add another column to the db to index them appropriately.  The standard API can only return one block for each slot so can't be used for non-canonical blocks.  Most of the time non-canonical blocks are just orphans but they can be completely different forks and so you wind up with multiple blocks at the same slot.
